src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMAlgorithm.java

```diff
@@ -0,0 +1,497 @@
+        logger.info(header);
+    }
+
+    /**
+     * A generic runner of a step in the EM algorithm
+     *
+     * @param routine a routine to run
+     * @param miscFactory a function from the routine signal to a miscellaneous iteration info string
+     * @param title title of the routine
+     * @param iterInfo the iteration info class (will be updated)
+     */
+    private void runRoutine(@Nonnull final Supplier<SubroutineSignal> routine,
+                            @Nonnull final Function<SubroutineSignal, String> miscFactory,
+                            @Nonnull final String title,
+                            @Nonnull EMAlgorithmIterationInfo iterInfo) {
+        final SubroutineSignal sig = routine.get();
+        final String misc = miscFactory.apply(sig);
+        iterInfo.errorNorm = sig.getDouble("error_norm");
+        iterInfo.logLikelihood = getLogLikelihood();
+        showIterationInfo(iterInfo.iter, title, iterInfo.logLikelihood, iterInfo.errorNorm, misc);
+    }
+
+    /**
+     * Run the full EM algorithm stack
+     *
+     * @return the exit status of the EM algorithm
+     */
+    public EMAlgorithmStatus runExpectationMaximization() {
+        params.validate();
+        if (params.adaptivePsiSolverModeSwitchingEnabled()) {
+            /* if copy ratio posterior calling is enabled, the first few iterations need to be robust.
+             * Therefore, we disable the target-resolved unexplained variance estimation (if enabled) and
+             * enforce isotropic unexplained variance */
+            if (params.adaptivePsiSolverModeSwitchingEnabled() &&
+                    !this.params.getPsiUpdateMode().equals(CoverageModelEMParams.PsiUpdateMode.PSI_ISOTROPIC)) {
+                this.params.setPsiPsiolverType(CoverageModelEMParams.PsiUpdateMode.PSI_ISOTROPIC);
+                logger.info("Overriding the requested unexplained variance solver to " +
+                        CoverageModelEMParams.PsiUpdateMode.PSI_ISOTROPIC.name());
+            }
+        }
+
+        showIterationHeader();
+
+        double prevEStepLikelihood = Double.NEGATIVE_INFINITY;
+        double latestEStepLikelihood = Double.NEGATIVE_INFINITY;
+        double prevMStepLikelihood = Double.NEGATIVE_INFINITY;
+        double latestMStepLikelihood = Double.NEGATIVE_INFINITY;
+        final EMAlgorithmIterationInfo iterInfo = new EMAlgorithmIterationInfo(Double.NEGATIVE_INFINITY, 0, 1);
+        boolean updateCopyRatioPosteriors = false;
+        boolean paramEstimationConverged = false;
+        boolean performMStep = true;
+
+        while (iterInfo.iter <= params.getMaxEMIterations()) {
+
+            /* cycle through E-step mean-field equations until they are satisfied to the desired degree */
+            double maxPosteriorErrorNorm = 0;
+            int iterEStep = 0;
+
+            while (iterEStep < params.getMaxEStepCycles()) {
+                double posteriorErrorNormReadDepth, posteriorErrorNormSampleUnexplainedVariance,
```

Comment:
It's just not done in Java, and breaking convention can make code harder to read.